app/dashboard/models.py

```diff
@@ -2400,6 +2430,136 @@ def is_staff(self):
         """
         return self.user.is_staff if self.user else False
 
+    def calc_activity_level(self):
+        """Determines the activity level of a user
+
+        Returns:
+            str: High, Low, Medium, or New
+
+        """
+        high_threshold = 15
+        med_threshold = 15
+        new_threshold_days = 7
+
+        if self.created_on > (timezone.now() - timezone.timedelta(days=new_threshold_days)):
+            return "New"
+
+        visits = self.actions.filter(action='Visit')
+        visits_last_month = visits.filter(created_on__gt=timezone.now() - timezone.timedelta(days=30)).count()
+
+        if visits_last_month > high_threshold:
+            return "High"
+        if visits_last_month > med_threshold:
+            return "Medium"
+        if visits_last_month > med_threshold:
+            return "Low"
+
+
+    def calc_longest_streak(self):
+        """ Determines the longest streak, in workdays, of this user
+
+        Returns:
+            int: a number of weekdays
+
+        """
+
+        # setup
+        action_dates = self.actions.all().values_list('created_on', flat=True)
+        start_date = timezone.datetime(self.created_on.year, self.created_on.month, self.created_on.day)
+        end_date = timezone.datetime(timezone.now().year, timezone.now().month, timezone.now().day)
+
+        # loop setup
+        iterdate = start_date
+        max_streak = 0
+        this_streak = 0
+        while iterdate < end_date:
+            # housekeeping
+            last_iterdate = start_date
+            iterdate += timezone.timedelta(days=1)
+
+            is_weekday = iterdate.weekday() < 5
+            if not is_weekday:
+                continue
+
+            has_action_during_window = len([ele for ele in action_dates if ele > last_iterdate.replace(tzinfo=pytz.utc) and ele < iterdate.replace(tzinfo=pytz.utc)])
+            if has_action_during_window:
+                this_streak += 1
+                if this_streak > max_streak:
+                    max_streak = this_streak
+            else:
+                this_streak = 0
+
+        return max_streak
+
+    def calc_num_repeated_relationships(self):
+        """ the number of repeat relationships that this user has created
+
+        Returns:
+            int: a number of repeat relationships
+
+        """
+        bounties = self.bounties
+        completed_bounties = bounties.filter(idx_status__in=['done'])
+        relationships = []
+        for bounty in completed_bounties:
+            fulfiller_handles = bounty.fulfillers_handles
+            if bounty.is_funder(self.handle):
+                for handle in fulfiller_handles:
+                    relationships.append(handle)
+            if self.handle in fulfiller_handles:
```

Comment:
nice this handles both funder -> fulfiller and and vice versa

Reply:
yar